nat-lab/tests/test_client_basic_stun.py

```diff
@@ -24,3 +29,71 @@ async def test_client_basic_stun(connection_tag: ConnectionTag, public_ip: str)
     async with new_connection_by_tag(connection_tag) as connection:
         ip = await testing.wait_long(stun.get(connection, config.STUN_SERVER))
         assert ip == public_ip, f"wrong public ip for the client {ip}"
+
+
+@pytest.mark.asyncio
+# Regression test for LLT-4306
+async def test_infinite_loop_in_stun_when_nonet() -> None:
+    async with AsyncExitStack() as exit_stack:
+        api = API()
+
+        (alpha, beta) = api.default_config_two_nodes()
+        alpha_connection = await exit_stack.enter_async_context(
+            new_connection_by_tag(ConnectionTag.DOCKER_CONE_CLIENT_1)
+        )
+        beta_connection = await exit_stack.enter_async_context(
+            new_connection_by_tag(ConnectionTag.DOCKER_CONE_CLIENT_2)
+        )
+
+        alpha_client = await exit_stack.enter_async_context(
+            Client(
+                alpha_connection,
+                alpha,
+                AdapterType.BoringTun,
+                telio_features=TelioFeatures(direct=Direct(providers=["stun"])),
+            ).run_meshnet(
+                api.get_meshmap(alpha.id),
+            )
+        )
+        beta_client = await exit_stack.enter_async_context(
+            Client(
+                beta_connection,
+                beta,
+                AdapterType.BoringTun,
+                telio_features=TelioFeatures(direct=Direct(providers=["stun"])),
+            ).run_meshnet(api.get_meshmap(beta.id))
+        )
+
+        await testing.wait_defined(
+            asyncio.gather(
+                alpha_client.wait_for_state_peer(
+                    beta.public_key, [State.Connected], [PathType.Direct]
+                ),
+                beta_client.wait_for_state_peer(
+                    alpha.public_key, [State.Connected], [PathType.Direct]
+                ),
+            ),
+            60,
+        )
+
+        for server in config.DERP_SERVERS:
+            await exit_stack.enter_async_context(
+                alpha_client.get_router().break_udp_conn_to_host(str(server["ipv4"]))
+            )
+            await exit_stack.enter_async_context(
+                beta_client.get_router().break_udp_conn_to_host(str(server["ipv4"]))
+            )
```

Comment:
Do we need to break connection for both clients?

Reply:
No we don't. I removed the break for `beta_client`